File: tunnel-client/src/main/java/com/theshuai/tunnelclient/peer/PeerIpPacket.java
package com.theshuai.tunnelclient.peer;

final class PeerIpPacket {
    private static final int IPV4_MIN_HEADER_BYTES = 20;
    private static final int IPV4_PROTOCOL_ICMP = 1;
    private static final int ICMP_ECHO_REPLY = 0;
    private static final int ICMP_ECHO_REQUEST = 8;

    private PeerIpPacket() {
    }

    static String destinationIpv4(byte[] packet) {
        if (packet == null || packet.length < IPV4_MIN_HEADER_BYTES) {
            return "";
        }
        int version = (packet[0] >>> 4) & 0x0F;
        int ihl = packet[0] & 0x0F;
        if (version != 4 || ihl < 5 || packet.length < ihl * 4) {
            return "";
        }
        return (packet[16] & 0xFF) + "."
                + (packet[17] & 0xFF) + "."
                + (packet[18] & 0xFF) + "."
                + (packet[19] & 0xFF);
    }

    static String sourceIpv4(byte[] packet) {
        if (packet == null || packet.length < IPV4_MIN_HEADER_BYTES) {
            return "";
        }
        int version = (packet[0] >>> 4) & 0x0F;
        int ihl = packet[0] & 0x0F;
        if (version != 4 || ihl < 5 || packet.length < ihl * 4) {
            return "";
        }
        return (packet[12] & 0xFF) + "."
                + (packet[13] & 0xFF) + "."
                + (packet[14] & 0xFF) + "."
                + (packet[15] & 0xFF);
    }

    static byte[] icmpEchoReplyFor(byte[] packet, String localVirtualIp) {
        if (!isIcmpEchoRequestFor(packet, localVirtualIp)) {
            return null;
        }
        int ihlBytes = (packet[0] & 0x0F) * 4;
        int totalLength = totalLength(packet);
        byte[] reply = new byte[totalLength];
        System.arraycopy(packet, 0, reply, 0, totalLength);

        for (int i = 0; i < 4; i++) {
            byte source = reply[12 + i];
            reply[12 + i] = reply[16 + i];
            reply[16 + i] = source;
        }
        reply[8] = 64;
        reply[10] = 0;
        reply[11] = 0;

        int icmpOffset = ihlBytes;
        reply[icmpOffset] = ICMP_ECHO_REPLY;
        reply[icmpOffset + 1] = 0;
        reply[icmpOffset + 2] = 0;
        reply[icmpOffset + 3] = 0;
        int icmpChecksum = checksum(reply, icmpOffset, totalLength - icmpOffset);
        reply[icmpOffset + 2] = (byte) ((icmpChecksum >>> 8) & 0xFF);
        reply[icmpOffset + 3] = (byte) (icmpChecksum & 0xFF);

        int ipChecksum = checksum(reply, 0, ihlBytes);
        reply[10] = (byte) ((ipChecksum >>> 8) & 0xFF);
        reply[11] = (byte) (ipChecksum & 0xFF);
        return reply;
    }

    private static boolean isIcmpEchoRequestFor(byte[] packet, String localVirtualIp) {
        if (packet == null || packet.length < IPV4_MIN_HEADER_BYTES || !destinationIpv4(packet).equals(localVirtualIp)) {
            return false;
        }
        int ihlBytes = (packet[0] & 0x0F) * 4;
        int totalLength = totalLength(packet);
        return totalLength >= ihlBytes + 8
                && totalLength <= packet.length
                && (packet[9] & 0xFF) == IPV4_PROTOCOL_ICMP
                && (packet[ihlBytes] & 0xFF) == ICMP_ECHO_REQUEST
                && (packet[ihlBytes + 1] & 0xFF) == 0;
    }

    private static int totalLength(byte[] packet) {
        if (packet == null || packet.length < 4) {
            return 0;
        }
        int totalLength = ((packet[2] & 0xFF) << 8) | (packet[3] & 0xFF);
        return totalLength > 0 && totalLength <= packet.length ? totalLength : packet.length;
    }

    private static int checksum(byte[] packet, int offset, int length) {
        long sum = 0;
        int index = offset;
        int end = offset + length;
        while (index + 1 < end) {
            sum += ((packet[index] & 0xFF) << 8) | (packet[index + 1] & 0xFF);
            index += 2;
        }
        if (index < end) {
            sum += (packet[index] & 0xFF) << 8;
        }
        while ((sum >>> 16) != 0) {
            sum = (sum & 0xFFFF) + (sum >>> 16);
        }
        return (int) (~sum) & 0xFFFF;
    }
}

File: implementations/csharp/client/src/Specus.Client/PeerMesh/PeerIpPacket.cs
using System.Buffers.Binary;
using System.Net;

namespace Specus.Client.PeerMesh;

internal static class PeerIpPacket
{
    private const byte ProtocolIcmp = 1;
    private const byte ProtocolTcp = 6;
    private const byte ProtocolUdp = 17;
    private const byte IcmpEchoReply = 0;
    private const byte IcmpEchoRequest = 8;
    private const byte IcmpDestinationUnreachable = 3;
    private const byte IcmpFragmentationNeeded = 4;

    public static string DestinationIPv4(ReadOnlySpan<byte> packet)
    {
        if (!IsIPv4(packet))
        {
            return "";
        }
        return new IPAddress(packet.Slice(16, 4)).ToString();
    }

    public static string SourceIPv4(ReadOnlySpan<byte> packet)
    {
        if (!IsIPv4(packet))
        {
            return "";
        }
        return new IPAddress(packet.Slice(12, 4)).ToString();
    }

    public static bool MatchesAuthenticatedEndpoints(ReadOnlySpan<byte> packet, string? peerVirtualIp,
        string? localVirtualIp) =>
        !string.IsNullOrWhiteSpace(peerVirtualIp)
        && !string.IsNullOrWhiteSpace(localVirtualIp)
        && string.Equals(SourceIPv4(packet), peerVirtualIp, StringComparison.Ordinal)
        && string.Equals(DestinationIPv4(packet), localVirtualIp, StringComparison.Ordinal);

    public static string FlowKey(ReadOnlySpan<byte> packet)
    {
        var source = SourceIPv4(packet);
        var target = DestinationIPv4(packet);
        if (source.Length == 0 || target.Length == 0)
        {
            return "";
        }
        var protocol = packet[9];
        var ihl = (packet[0] & 0x0f) * 4;
        var sourcePort = 0;
        var targetPort = 0;
        if ((protocol is ProtocolTcp or ProtocolUdp) && packet.Length >= ihl + 4)
        {
            sourcePort = BinaryPrimitives.ReadUInt16BigEndian(packet.Slice(ihl, 2));
            targetPort = BinaryPrimitives.ReadUInt16BigEndian(packet.Slice(ihl + 2, 2));
        }
        return $"{source}:{sourcePort}->{target}:{targetPort}/{protocol}";
    }

    public static byte[]? IcmpEchoReplyFor(ReadOnlySpan<byte> packet, string? localVirtualIp)
    {
        if (!IsIcmpEchoRequestFor(packet, localVirtualIp))
        {
            return null;
        }
        var ihl = (packet[0] & 0x0f) * 4;
        var totalLength = TotalLength(packet);
        var reply = packet[..totalLength].ToArray();
        reply.AsSpan(16, 4).CopyTo(reply.AsSpan(12, 4));
        packet.Slice(12, 4).CopyTo(reply.AsSpan(16, 4));
        reply[8] = 64;
        reply[10] = 0;
        reply[11] = 0;

        reply[ihl] = IcmpEchoReply;
        reply[ihl + 1] = 0;
        reply[ihl + 2] = 0;
        reply[ihl + 3] = 0;
        BinaryPrimitives.WriteUInt16BigEndian(reply.AsSpan(ihl + 2, 2), Checksum(reply.AsSpan(ihl, totalLength - ihl)));
        BinaryPrimitives.WriteUInt16BigEndian(reply.AsSpan(10, 2), Checksum(reply.AsSpan(0, ihl)));
        return reply;
    }

    public static byte[] ClampTcpMss(byte[] packet, int pathMtu)
    {
        ReadOnlySpan<byte> view = packet;
        if (!IsIPv4(view) || view[9] != ProtocolTcp)
        {
            return packet;
        }
        var ipHeaderLength = (view[0] & 0x0f) * 4;
        var totalLength = TotalLength(view);
        if (totalLength < ipHeaderLength + 20 || (view[ipHeaderLength + 13] & 0x02) == 0)
        {
            return packet;
        }
        var tcpHeaderLength = (view[ipHeaderLength + 12] >> 4) * 4;
        if (tcpHeaderLength < 20 || totalLength < ipHeaderLength + tcpHeaderLength)
        {
            return packet;
        }
        var maxMss = Math.Max(536, pathMtu - ipHeaderLength - 20);
        for (var cursor = ipHeaderLength + 20; cursor < ipHeaderLength + tcpHeaderLength;)
        {
            var kind = view[cursor];
            if (kind == 0)
            {
                break;
            }
            if (kind == 1)
            {
                cursor++;
                continue;
            }
            if (cursor + 1 >= ipHeaderLength + tcpHeaderLength)
            {
                break;
            }
            var optionLength = view[cursor + 1];
            if (optionLength < 2 || cursor + optionLength > ipHeaderLength + tcpHeaderLength)
            {
                break;
            }
            if (kind == 2 && optionLength == 4)
            {
                var advertised = BinaryPrimitives.ReadUInt16BigEndian(view.Slice(cursor + 2, 2));
                if (advertised <= maxMss)
                {
                    return packet;
                }
                var clamped = packet.ToArray();
                BinaryPrimitives.WriteUInt16BigEndian(clamped.AsSpan(cursor + 2, 2), checked((ushort)maxMss));
                clamped[ipHeaderLength + 16] = 0;
                clamped[ipHeaderLength + 17] = 0;
                BinaryPrimitives.WriteUInt16BigEndian(
                    clamped.AsSpan(ipHeaderLength + 16, 2),
                    TcpChecksum(clamped, ipHeaderLength, totalLength - ipHeaderLength));
                return clamped;
            }
            cursor += optionLength;
        }
        return packet;
    }

    public static byte[]? IcmpFragmentationNeededFor(ReadOnlySpan<byte> packet, int pathMtu)
    {
        if (!IsIPv4(packet))
        {
            return null;
        }
        var originalHeaderLength = (packet[0] & 0x0f) * 4;
        var originalLength = TotalLength(packet);
        var quotedLength = Math.Min(originalLength, originalHeaderLength + 8);
        var response = new byte[20 + 8 + quotedLength];
        response[0] = 0x45;
        BinaryPrimitives.WriteUInt16BigEndian(response.AsSpan(2, 2), checked((ushort)response.Length));
        response[8] = 64;
        response[9] = ProtocolIcmp;
        packet.Slice(16, 4).CopyTo(response.AsSpan(12, 4));
        packet.Slice(12, 4).CopyTo(response.AsSpan(16, 4));
        response[20] = IcmpDestinationUnreachable;
        response[21] = IcmpFragmentationNeeded;
        BinaryPrimitives.WriteUInt16BigEndian(response.AsSpan(26, 2), checked((ushort)pathMtu));
        packet[..quotedLength].CopyTo(response.AsSpan(28));
        BinaryPrimitives.WriteUInt16BigEndian(response.AsSpan(22, 2), Checksum(response.AsSpan(20)));
        BinaryPrimitives.WriteUInt16BigEndian(response.AsSpan(10, 2), Checksum(response.AsSpan(0, 20)));
        return response;
    }

    private static bool IsIPv4(ReadOnlySpan<byte> packet)
    {
        if (packet.Length < 20 || packet[0] >> 4 != 4)
        {
            return false;
        }
        var ihl = (packet[0] & 0x0f) * 4;
        return ihl >= 20 && packet.Length >= ihl;
    }

    private static bool IsIcmpEchoRequestFor(ReadOnlySpan<byte> packet, string? localVirtualIp)
    {
        if (DestinationIPv4(packet) != localVirtualIp)
        {
            return false;
        }
        var ihl = (packet[0] & 0x0f) * 4;
        var totalLength = TotalLength(packet);
        return totalLength >= ihl + 8
            && totalLength <= packet.Length
            && packet[9] == ProtocolIcmp
            && packet[ihl] == IcmpEchoRequest
            && packet[ihl + 1] == 0;
    }

    private static int TotalLength(ReadOnlySpan<byte> packet)
    {
        if (packet.Length < 4)
        {
            return 0;
        }
        var totalLength = BinaryPrimitives.ReadUInt16BigEndian(packet.Slice(2, 2));
        return totalLength > 0 && totalLength <= packet.Length ? totalLength : packet.Length;
    }

    internal static ushort Checksum(ReadOnlySpan<byte> data)
    {
        uint sum = 0;
        var index = 0;
        while (index + 1 < data.Length)
        {
            sum += BinaryPrimitives.ReadUInt16BigEndian(data.Slice(index, 2));
            index += 2;
        }
        if (index < data.Length)
        {
            sum += (uint)data[index] << 8;
        }
        while ((sum >> 16) != 0)
        {
            sum = (sum & 0xffff) + (sum >> 16);
        }
        return (ushort)~sum;
    }

    private static ushort TcpChecksum(ReadOnlySpan<byte> packet, int tcpOffset, int tcpLength)
    {
        uint sum = 0;
        for (var index = 12; index < 20; index += 2)
        {
            sum += BinaryPrimitives.ReadUInt16BigEndian(packet.Slice(index, 2));
        }
        sum += ProtocolTcp;
        sum += checked((uint)tcpLength);
        var segment = packet.Slice(tcpOffset, tcpLength);
        var cursor = 0;
        while (cursor + 1 < segment.Length)
        {
            sum += BinaryPrimitives.ReadUInt16BigEndian(segment.Slice(cursor, 2));
            cursor += 2;
        }
        if (cursor < segment.Length)
        {
            sum += (uint)segment[cursor] << 8;
        }
        while ((sum >> 16) != 0)
        {
            sum = (sum & 0xffff) + (sum >> 16);
        }
        return (ushort)~sum;
    }
}
